### PR TITLE
Practice5: todo list

### PR DESCRIPTION
**Pull request by LaxTheMan**
# Task
[Practice 5](https://github.com/pj100-team/react-v3/issues/5)

#Changes made
- Created a todo list with an add, delete, selected delete functionality
- Added checkbox selection for deleting selectively or bulk delete

#Demo video

https://github.com/user-attachments/assets/019bcb8d-2290-4e3c-8d6a-6160ae9fb9d4

### CODE REVIEW

File: src/components/TODOList.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import React, { useState } from 'react';
+import { Todo } from '../templates/Practice5';
+
+type TODOListProps = {
+  todos: Todo[];
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
+};
+
+export const TODOList = ({ todos, setTodos }: TODOListProps) => {
+  const [allChecked, setAllChecked] = useState<boolean>(false);
```
Comment (ThanainanLi) [inline src/components/TODOList.tsx]: To Follow our [React guidelines ](https://pj100.esa.io/posts/16722#%E3%82%A2%E3%83%B3%E3%83%81%E3%83%91%E3%82%BF%E3%83%BC%E3%83%B3%E3%81%AE%E6%94%B9%E5%96%84)

```suggestion
  const [isAllChecked, setIsAllChecked] = useState<boolean>(false);
```

File: src/templates/Practice4.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import React from 'react';
+import { useForm, SubmitHandler } from 'react-hook-form';
+import { InputField } from '../components/InputField';
+import { useState } from 'react';
+
+export type FormInput = {
+  'Postal Code': string;
+  Prefecture: string;
+  'City/Ward/Town': string;
```
Comment (ThanainanLi) [inline src/templates/Practice4.tsx]: For consistency, I think we should ensure the key types are the same.

```suggestion
  Postal Code: string;
  Prefecture: string;
  CityWardTown: string;
```

**Review by ThanainanLi — COMMENTED** [review]
Thank you for your PR. Few comments from me!

File: src/components/TODOList.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                  </button>
+                </td>
+              </tr>
+            ))}
+          </tbody>
+        </table>
+      )}
+    </>
+  );
+};
```
Comment (extramix) [inline src/components/TODOList.tsx]: The word "to-do" in "to-do list" is not an acronym so it shouldn't be all uppercases. I think this file should be named `TodoList.tsx`

File: src/components/TODOList.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  };
+
+  const handleDeleteSelected = () => {
+    setAllChecked(false);
+    setTodos(todos.filter((todo) => todo.checked === false));
+  };
+
+  const handleCheckBoxClick = (id: number) => {
+    setTodos(todos.map((todo) => (todo.id === id ? { ...todo, checked: !todo.checked } : todo)));
+  };
```
Comment (extramix) [inline src/components/TODOList.tsx]: Unless this callback is directly passed onto the event handler (i.e.`onChange={handleChageOfSomething}`), The callback name should **reflect on what it does and not where it is handled.** 

That way, other people can understand the purpose of this right away without having to interpret what's inside.


How about `toggleTodoCheckedStatus`?

File: src/templates/Practice4.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            onClick={handleClick}
+          >
+            Search
+          </button>
+        </div>
+        <InputField
+          label="Prefecture"
+          register={register}
+          validationRules={{
+            required: '必須項目です',
```
Comment (extramix) [inline src/templates/Practice4.tsx]: Please put constant error messages in a separate file!

**Review by extramix — CHANGES_REQUESTED** [review]

File: src/components/TODOList.tsx (hunk 1 above)
Comment (LaxTheMan) [reply src/components/TODOList.tsx]: Changed state name to isAllChecked

File: src/components/TODOList.tsx (hunk 4 above)
Comment (LaxTheMan) [reply src/components/TODOList.tsx]: Understood! Changed event handler name to toggleTodoCheckedStatus

File: src/components/TODOList.tsx (hunk 3 above)
Comment (LaxTheMan) [reply src/components/TODOList.tsx]: Changed filename to `TodoList.tsx`

File: src/templates/Practice4.tsx (hunk 2 above)
Comment (LaxTheMan) [reply src/templates/Practice4.tsx]: Already fixed in a previous commit, but commits did not show. Fixed key types are showing now!

File: src/templates/Practice4.tsx (hunk 5 above)
Comment (LaxTheMan) [reply src/templates/Practice4.tsx]: Already fixed in a previous commit, but commits did not show. Error messages have been separated into `messages.ts` file!

**Review by LaxTheMan — COMMENTED** [review]
Fixed issues

**Review by extramix — APPROVED** [review]
LGTM!

**Review by ThanainanLi — APPROVED** [review]
LGTM!

**Pull request merged**